src/Microsoft.NetCore.Analyzers/UnitTests/Tasks/UseValueTaskCorrectlyTests.cs

```diff
@@ -0,0 +1,1274 @@
+                        GetCSharpResultAt(10, 30, UseValueTasksCorrectlyAnalyzer.AccessingIncompleteResultRule),
+                        // Test0.cs(17,30): warning CA2012: Use ValueTasks correctly.
+                        GetCSharpResultAt(17, 30, UseValueTasksCorrectlyAnalyzer.AccessingIncompleteResultRule),
+                        // Test0.cs(24,30): warning CA2012: Use ValueTasks correctly.
+                        GetCSharpResultAt(24, 30, UseValueTasksCorrectlyAnalyzer.AccessingIncompleteResultRule)
+                    });
+        }
+
+        [Theory]
+        [MemberData(nameof(IsCompleteProperties))]
+        public void Diagnostics_AssertsLaterInFlowAfterDirectLocalAccess(string isProp)
+        {
+            VerifyCSharp(CSharpBoilerplate(@"
+                using System;
+                using System.Diagnostics;
+                using System.Threading.Tasks;
+
+                class C
+                {
+                    public void AssertsLaterInFlowAfterDirectAccess()
+                    {
+                        ValueTask vt = Helpers.ReturnsValueTask();
+                        if (DateTime.Now > DateTime.Now)
+                        {
+                            vt.GetAwaiter().GetResult();
+                        }
+                        else
+                        {
+                            vt.GetAwaiter().GetResult();
+                        }
+                        Debug.Assert(vt." + isProp + @");
+                    }
+
+                    public void AssertsLaterInFlowAfterDirectAccessOfT<T>()
+                    {
+                        ValueTask<T> vt = Helpers.ReturnsValueTaskOfT<T>();
+                        if (DateTime.Now > DateTime.Now)
+                        {
+                            vt.GetAwaiter().GetResult();
+                        }
+                        else
+                        {
+                            vt.GetAwaiter().GetResult();
+                        }
+                        Debug.Assert(vt." + isProp + @");
+                    }
+
+                    public void AssertsLaterInFlowAfterDirectAccessOfTResult<T>()
+                    {
+                        ValueTask<T> vt = Helpers.ReturnsValueTaskOfT<T>();
+                        if (DateTime.Now > DateTime.Now)
+                        {
+                            T result = vt.Result;
+                        }
+                        else
+                        {
+                            T result = vt.Result;
+                        }
+                        Debug.Assert(vt." + isProp + @");
+                    }
+                }"),
+                    parseOptions: s_csharpPreviewParseOptions,
+                    expected: new DiagnosticResult[]
+                    {
+                        // Test0.cs(10,40): warning CA2012: Use ValueTasks correctly.
+                        GetCSharpResultAt(10, 40, UseValueTasksCorrectlyAnalyzer.AccessingIncompleteResultRule),
```

Comment:
With the new test framework, you can actually add annotations to original source code for the diagnostic location, which makes it easier to understand the squiggle span. I am presuming this diagnostic is flagging the value task creation and not the `.GetResult()` or `.Result` operations. Is that correct? If so, should the diagnostic instead be flagging the incorrect use site. i.e. should we have 6 diagnostics here, 2 in each method?

Reply:
> If so, should the diagnostic instead be flagging the incorrect use site. i.e. should we have 6 diagnostics here, 2 in each method?

It's flagging the first usage failure for each creation site; as soon as it hits a failure, it stops looking for others.